client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx

```diff
@@ -18,140 +16,157 @@ import {
 
 export const SearchWithUserSource = (
   props: ControlProps & { options: UserOptions }
 ) => {
-  const { data, path, handleChange, label, visible, options } = props;
+  const {
+    data,
+    path,
+    handleChange,
+    label,
+    visible,
+    options,
+    schema,
+    renderers,
+  } = props;
   const t = useTranslation('programs');
-  const [searchText, setSearchText] = useState('');
-  const [editMode, setEditMode] = useState(!data);
-  const [noResultsText, setNoResultsText] = useState(
-    t('control.search.searching-label')
-  );
+
+  const isPatientSelected = !!data?.id;
 
   const {
     runQuery,
     getOptionLabel,
-    getDisplayElement,
     saveFields,
-    placeholderText,
-    loading,
     error: queryError,
     results,
+    resetResults,
   } = useSearchQueries(options ?? {});
 
-  const debouncedOnChange = useDebounceCallback(
-    value => {
-      if (value.length >= MIN_CHARS) runQuery(value);
-      else {
-        // Clear the results if user input falls *below* `minChars`
-        if (results.length) runQuery('');
-      }
-      setNoResultsText(t('control.search.no-results-label'));
-    },
-    [searchText],
-    500
-  );
-
-  const handleDataUpdate = (selectedResult: Record<string, unknown> | null) => {
-    if (selectedResult === null) return;
-    if (!saveFields) handleChange(path, selectedResult);
-    else {
-      const newObj: Record<string, unknown> = {};
-      saveFields?.forEach(
-        field => (newObj[field] = selectedResult[field] ?? null)
-      );
-      handleChange(path, newObj);
+  useEffect(() => {
+    if (!data) {
+      handleChange(path, {});
+      return;
     }
-    setEditMode(false);
-  };
 
-  const getNoOptionsText = () => {
-    switch (true) {
-      case loading:
-        return t('control.search.searching-label');
-      case searchText.length < MIN_CHARS:
-        return t('control.search.below-min-chars', { minChars: MIN_CHARS });
-      default:
-        return noResultsText;
+    // Only run the database query if a specific patient hasn't yet been
+    // selected
+    if (isPatientSelected) return;
+
+    const searchFilter: FilterBy = {};
+    options.searchFields.forEach(field => {
+      const match =
+        field in queryMatchTypes
+          ? queryMatchTypes?.[field as keyof typeof queryMatchTypes]
+          : 'like';
+      if (data[field]) searchFilter[field] = { [match]: data[field] };
+    });
+    if (Object.keys(searchFilter).length > 0) runQuery(searchFilter);
+    else resetResults();
+  }, [data]);
+
+  const handlePatientSelect = (patientId: string) => {
+    const patient = results.find(p => (p.id = patientId));
+    if (!patient) return;
+    if (!saveFields) {
+      handleChange(path, patient);
+      return;
     }
+    const newData = Object.fromEntries(
+      Object.entries(patient).filter(
+        ([key]) => (saveFields as string[])?.includes(key)
+      )
+    );
+    handleChange(path, newData);
```

Comment:
Do we want to update the local fields or just display the queried data from the selected patient? would have left the local data untouched, for example, when resetting the found patient all previously entered data is cleared which could be a bit annoying. But maybe not really important @richardmoizeau?

Reply:
I considered that, but I thought it was not worth the trouble for now. Most of the time you're going to select the correct patient the first time, and in the occasional event you don't, you're probably going to find them again with only a small amount of input (1-2 fields).

I think it could be annoying if you need to enter data manually (i.e. not in database), and end up selecting a match anyway. Then when you "reset" it, you lose your data. However, this is the kind of mistake people will only make once or twice -- nothing like having to re-do your form filling for people to learn from their mistakes. (i.e. I don't think it'll happen to often).

If this proves to be a pain point, we can address it at a later stage.

Reply:
:-) not sure if that the kind of UX we should be aiming for. Think would be good to go through the control with @anisha-msupply anyway, e.g. also not 100% if the unlink button can't be handled in a nicer way, e.g. with some extra visual hints that the linked patient is an mSupply patient...